File: backend/src/schemas/entrada_estoque.py
"""
Schemas Pydantic para Entrada de Estoque (multi-produto).
"""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class EntradaEstoqueCreate(BaseModel):
    """Schema para criar entrada."""
    tipo_produto: str = Field(..., description="Tipo: oleo, filtro, peca")
    produto_id: int = Field(..., description="ID do produto")
    quantidade_litros: Decimal = Field(..., gt=0, description="Quantidade")
    custo_unitario: Decimal = Field(..., ge=0, description="Custo unitário")
    fornecedor: str | None = Field(None, max_length=100, description="Fornecedor")
    nota_fiscal: str | None = Field(None, max_length=50, description="Número da NF")
    data_compra: date = Field(..., description="Data da compra")
    observacoes: str | None = Field(None, description="Observações")

    @field_validator("tipo_produto")
    @classmethod
    def validar_tipo(cls, v: str) -> str:
        if v not in ("oleo", "filtro", "peca"):
            raise ValueError("Tipo deve ser 'oleo', 'filtro' ou 'peca'")
        return v


class EntradaEstoqueResponse(BaseModel):
    """Schema de resposta."""
    id: int
    tipo_produto: str
    produto_id: int
    produto_nome: str
    produto_marca: str
    quantidade_litros: Decimal
    custo_unitario: Decimal
    custo_total: Decimal
    fornecedor: str | None
    nota_fiscal: str | None
    data_compra: date
    observacoes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EntradaEstoqueListResponse(BaseModel):
    """Resposta paginada."""
    items: list[EntradaEstoqueResponse]
    total: int
    page: int
    pages: int


class ProdutoBuscaResponse(BaseModel):
    """Resultado de busca de produto."""
    tipo: str = Field(description="oleo, filtro, peca")
    id: int
    codigo_produto: str | None = None
    nome: str
    marca: str
    label: str = Field(description="Texto formatado para exibição")

File: frontend/src/types/entrada_estoque.ts
export interface EntradaEstoque {
  id: number
  tipo_produto: string
  produto_id: number
  produto_nome: string
  produto_marca: string
  quantidade_litros: string
  custo_unitario: string
  custo_total: string
  fornecedor: string | null
  nota_fiscal: string | null
  data_compra: string
  observacoes: string | null
  created_at: string
  updated_at: string
}

export interface EntradaEstoqueCreate {
  tipo_produto: string
  produto_id: number
  quantidade_litros: number
  custo_unitario: number
  fornecedor?: string | null
  nota_fiscal?: string | null
  data_compra: string
  observacoes?: string | null
}

export interface EntradaEstoqueListResponse {
  items: EntradaEstoque[]
  total: number
  page: number
  pages: number
}

export interface ProdutoBusca {
  tipo: string
  id: number
  codigo_produto: string | null
  nome: string
  marca: string
  label: string
}
